---
title: "MetaPhlAn4 Processing and Cleaning"
output: html_notebook
editor_options: 
  chunk_output_type: inline
---


Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

This markdown will process short read sequences in fastq.gz format and output taxonomic
profiles from MetaPhlAn4. 
GitHub: https://github.com/biobakery/MetaPhlAn
Publication: https://www.biorxiv.org/content/10.1101/2022.08.22.504593v1 

Set up software for running MetaPhlAn4 in linux

1. Install Anaconda or miniconda (a smaller version that is sufficient for 
MetaPhlAn4 but if frequently running sequence analysis and multiple tools
it may be worth it to use Anaconda)
Anaconda: https://docs.anaconda.com/anaconda/install/
miniconda: https://docs.anaconda.com/miniconda/ 

2. Install MetaPhlAn4 
GitHub link: https://github.com/biobakery/MetaPhlAn/wiki/MetaPhlAn-4#installation
Supplemental help: https://omicx.cc/posts/2021-12-17-install-and-setup-metaphlan-3/

$ conda install -c bioconda metaphlan

3. Troubleshooting
If you can't move ahead yet, try adding channels for conda-forge and bioconda and 
ordering channels:

	$ conda config --add channels defaults
  $ conda config --add channels bioconda
  $ conda config --add channels conda-forge
  
  $ conda config –set channel_priority flexible
  $ conda create –name mpa –c conda-forge –c bioconda metaphlan
  
  Install bowtie inside this environment
  {add link for installation}
  $ metaphlan --install --bowtie2db <database folder>
  

This will install the package and CHOCOPhlAn database. 

Running MetaPhlAn

1. Activate MetaPhlAn environment within conda
  
  $ conda activate mpa
  
2. Set directory to folder containing sequences (input folder)

If stored locally:

  $ cd /mnt/c/User/pennkey/Downloads/sequence_folder
  
If stored in box sync (for large quantities of files)

  $ cd /mnt/c/User/pennkey/Box\ Sync/sequence_folder
  
3. Set up your output files

Choose where to locate your output files (locally, GitHub, or Box Sync)
If running remotely, Box Sync is a good option as you can continue to monitor 
progress and access data externally without logging in. 

Create 2 output files within your 'metaphlan' output file:
1. profiles
2. bowtie2

The profiles will be text or tsv files containing taxonomic assignments and read
counts. 


4. Test your run on an individual sample

To run individual (non read-paired samples):
  pennkey/input/directory $ metaphlan file_1.fastq.gz --input_type fastq 
                            -o .../metaphlan/profiles/profiled_metagenome.txt (or .tsv)
                            --bowtie2out .../metaphlan/bowtie2/metagenome.bowtie2.bz2
                            
To run on read-paired data (seq_1_R1.fastq, seq_1_R2.fastq)

  pennkey/input/directory $ metaphlan file_1_R1.fastq.gz, file_1_R2.fastq.gz                                       --bowtie2out                                                                         .../metaphlan/bowtie2/file_1_metagenome.bowtie2.bz2 
                             --nproc 5 –input_type fastq 
                          –o .../metaphlan/profiles/file_1_profiled_metagenome.txt

5. To run a loop on all sequences in a folder, set the directory in that folder. 
Ex.samples are labelled UM1_R1.fastq.gz and UM1_R2.fastq.gz

  pennkey/input/directory $ for i in *_R1.fastq.gzl do metaphlan
                            $i,${i/_R1.fastq.gz/_R2.fastq.gz}
                            --input_type fastq -nproc 8
                            --bowtie2out .../bowtie2/${i%}.bowtie2.bz2
                            -o .../profiles/${i%}_profiled.txt ; done
                            
6. Example using paired read fasta files and explanations (example reading from
and into box)
  For this example, files are stored in box sync folder titled "project_1" and
  sample ID's are S1, S2, S3... and read-paired
    a. Sync box files to desktop
    b. Create new folders: metaphlan/bowtie2 and metaphlan/profiles
    
    $ conda activate mpa
    $ cd /mnt/c/User/Box\ Sync/project_1
    .../Box\ Sync/project_1/ $ for i in *R1.fastq.gz; #for files with ID.fastq
                                *perform metaphlan on both R1 and R2 sequences per ID
                                *and pull the string before "_R*.fastq.gz" as the file                                 *ID    
                                do metaphlan $i,{i/_R1.fastq.gz/_R2.fastq.gz} 
                                #Read all fastq files in the folder
                                --input_type fastq
                                #Run on 8 cores (can change- storage and speed)
                                -nproc 8 
                                #bowtie files will output to the bowtie file with
                                #"ID".bowtie2.bz2
                                --bowtie2out .../project_1/metaphlan/bowtie2/                                         ${i%}.bowtie2.bz2
                                #profiles will output to profile folder as
                                # "ID"_profile.txt
                                -o .../project_1/metaphlan/profiles/
                                ${i%}_profile.txt 
                                #run for each sequence and end loop
                                ; done
     
     All together:
     For i in *_R1.fastq.gz; do metaphlan $i.${i/_R1.fastq.gz/_R2.fastq.gz} --             input_type fastq –nproc 10 –bowtie2out ../../project name metaphlan                   output/bowtie2/${i%}.bowtie2.bz2 -o ../../project name metaphlan output/              profiles/${i%}_profiled.txt; done      
     
     
     
     
  Metaphlan output structure and cleaning
 Metaphlan outputs individual text files for each sample with read counts and
 taxonomic profiles (Kingdom....Strain). Each file will be named with the sample
 ID which must be pulled from the listfile to create one datafile combining all
 sample profiles. 
 
                                
Set up
```{r}
library(tidyverse)
library(janitor)
library(gt)
library(ggplot2)
library(viridis)
```


Set working directory to output folder (ex. project_1/data/metaphlan/profiles)

```{r}
list.files("data/metaphlan/profiles") |>
  tibble(file_name = _) |>
  mutate(specimen_name = file_name) |>
  mutate(specimen_name = gsub("_R1_001.fastq.gz_profiled.txt", "", specimen_name)) |>
  mutate(long_file_name = list.files("data/metaphlan/profiles", full.names = TRUE)) |>
  mutate(mpa_dat = map(.x = long_file_name, .f = ~ read_tsv(.x, skip = 4) |> mutate_all(.funs = ~ as.character(.x)))) |>
  select(specimen_name, mpa_dat) |>
  unnest(cols = c(mpa_dat)) |>
  mutate(relative_abundance = as.numeric(relative_abundance) / 100) |>
  identity() -> all_mpa_files
all_mpa_files
```


 Some classifications only reach a certain
 level of specificity, e.g. Genus, and will therefore output "unknown" for the 
 rest of the class levels (species, strain). Hashed code will classify these
 cases as such. 

```{r}
all_mpa_files |>
  rename_all(.funs = ~ gsub("#","", .x)) |>
  janitor::clean_names() |>
  mutate(clade_name = replace(clade_name, clade_name == "UNCLASSIFIED", "t__UNCLASSIFIED"))|>
  filter(grepl("t__",clade_name))|>
  select(-additional_species) |>
  separate_wider_delim(clade_name,"|", names = c("Kingdom","Phylum","Class","Order","Family","Genus","Species","Strain"), too_few = "align_start") |>
  select(-ncbi_tax_id) |>
  rename(specimen_id = specimen_name) |>
  mutate(Kingdom = gsub("k__|t__", "", Kingdom)) |>
  mutate(Phylum = gsub("p__", "", Phylum)) |>
  mutate(Class = gsub("c__", "", Class)) |>
  mutate(Order = gsub("o__", "", Order)) |>
  mutate(Family = gsub("f__", "", Family)) |>
  mutate(Genus = gsub("g__", "", Genus)) |>
  mutate(Species = gsub("s__", "", Species)) |>
  mutate(Strain = gsub("t__", "", Strain)) |>
  #Can add this code to above block to classify alphanumeric outputs as 'Unknown'
# mutate(Class = case_when(str_detect(Class, "CFGB") ~ "Unknown", .default = as.character(Class))) |>
# mutate(Order = case_when(str_detect(Order, "OFGB") ~ "Unknown", .default = as.character(Order))) |>
# mutate(Family = case_when(str_detect(Family, "FGB") ~ "Unknown", .default = as.character(Family))) |>
# mutate(Genus = case_when(str_detect(Genus, "GGB") ~ "Unknown", .default = as.character(Genus))) |>
# mutate(Species = case_when(str_detect(Species, "SGB") ~ "Unknown", .default = as.character(Species))) |>
  identity() -> all_mpa_cleaned
all_mpa_cleaned

write_csv(all_mpa_cleaned, "data/metaphlan_profiles_cleaned.csv")
```


Number of reads processed

```{r}
list.files("data/metaphlan/profiles") |>
  tibble(file_name = _) |>
  mutate(specimen_name = file_name) |>
  mutate(specimen_name = gsub("_R1_001.fastq.gz_profiled.txt", "", specimen_name)) |>
  mutate(long_file_name = list.files("data/metaphlan/profiles", full.names = TRUE)) |> 
  mutate(mpa_dat = map(.x = long_file_name, .f = ~ read_lines(.x, skip = 2, n_max = 1))) |>
  select(specimen_name, mpa_dat) |>
  unnest(cols = c(mpa_dat)) |>
  mutate(reads_processed = as.numeric(gsub("#| reads processed","", mpa_dat))) |> 
  select(-mpa_dat) |> 
  identity() -> mpa_reads_processed
mpa_reads_processed
```
 
 
 Quick visualization of data
 
```{r}
#stacked barplot
all_mpa_cleaned |>
  select(Family, relative_abundance,specimen_id) |>
  mutate(Family = gsub("f__", "", Family)) |>
  mutate(Family = case_when(str_detect(Family, "FGB") ~ "Unknown", .default = as.character(Family))) |>
  ggplot(aes(fill = Family, y = relative_abundance, x = specimen_id)) +
  geom_bar(position = "fill", stat = "identity") +
  theme(legend.position = "bottom") +
  scale_fill_viridis(discrete = T) +
  theme(axis.text.x = ggtext::element_markdown(color = "black", angle = 45, hjust = 1, vjust = 1),
        axis.text.y = ggtext::element_markdown(color = "black"),
        legend.title = ggtext::element_markdown(color = "black"),
        legend.text = element_text(size = 5)) +
#  guides(color = guide_legend(override.aes = list(size = 0.5))) +
  labs(x = "Specimen ID", fill = "Bacterial Family", y = "Relative Abundance") |>
  identity() -> music_metaphlan_family_composition

ggsave(music_metaphlan_family_composition, filename = "figs/music_metaphlan_family_comp.png", width = 23, height = 30, units = "cm")
```

